### ISSUE TITLE
Capture key name in reader exception where applicable

### ISSUE THREAD

**Issue by momania** [plokhotnyuk/jsoniter-scala#433]
It would be very helpful to be able to generate a tiny bit more user friendly error messages that include the key in case of a value parsing error, so it's easier for myself, but also an enduser to correct any input errors.

i.e. when a `LocalDate` has a wrong format, the following exception message is thrown:
`expected digit, offset: 0x0000012d`

Looking at the macro generated code:
```scala
case 752419634 => if (in.isCharBufEqualsTo(l, "tradeDate"))
  {
    if (p0.&(65536).!=(0))
      p0.^=(65536)
    else
      in.duplicatedKeyError(l);
    _tradeDate = in.readLocalDate(_tradeDate)
}
```
I think there is an opportunity to wrap the last call (or the whole block that's in the while loop) in a separate try/catch and enrich the thrown exception with the key name (`tradeDate` in this case).

The fact that the user knows which key has an error usually is enough. The error message itself can stay technical.

**Comment by plokhotnyuk** [comment]
@momania by default jsoniter-scala prints a hex-dump of the buffer where an error is happening, like:
```
expected digit, offset: 0x0000004d, buf:
+----------+-------------------------------------------------+------------------+
|          |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f | 0123456789abcdef |
+----------+-------------------------------------------------+------------------+
| 00000020 | 69 22 3a 22 32 30 30 37 2d 31 32 2d 30 33 54 31 | i":"2007-12-03T1 |
| 00000030 | 30 3a 31 35 3a 33 30 2e 30 30 31 5a 22 2c 22 6c | 0:15:30.001Z","l |
| 00000040 | 64 22 3a 22 32 30 30 37 2d 31 32 2d 33 22 2c 22 | d":"2007-12-3"," |
| 00000050 | 6c 64 74 22 3a 22 32 30 30 37 2d 31 32 2d 30 33 | ldt":"2007-12-03 |
| 00000060 | 54 31 30 3a 31 35 3a 33 30 22 2c 22 6c 74 22 3a | T10:15:30","lt": |
+----------+-------------------------------------------------+------------------+```

So you can easily spot the key and surrounding context. Will it be acceptable solution for your case?

**Comment by momania** [comment]
I know that collecting the hex-dump is the default behaviour and for me as the developer that's fine to see in my logs. 
But for my API end user I'd like to be able to construct a more meaningful error message, and such a hex dump doesn't do well in a json formatted error response. Just the format itself is already an issue if they look at the raw response. 

i.e.
```
{
    "error": "Unable to parse input body",
    "reason": "expected digit, offset: 0x0000012d, buf:\n+----------+-------------------------------------------------+------------------+\n|          |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f | 0123456789abcdef |\n+----------+-------------------------------------------------+------------------+\n| 00000100 | 72 72 65 6e 63 79 22 3a 20 22 55 53 44 22 2c 0a | rrency\": \"USD\",. |\n| 00000110 | 0a 20 20 20 20 20 20 20 20 20 20 20 20 22 74 72 | .            \"tr |\n| 00000120 | 61 64 65 44 61 74 65 22 3a 20 22 32 32 20 4e 6f | adeDate\": \"22 No |\n| 00000130 | 76 20 32 30 31 39 22 2c 0a 0a 20 20 20 20 20 20 | v 2019\",..       |\n| 00000140 | 20 20 20 20 20 20 22 65 78 70 69 72 79 44 61 74 |       \"expiryDat |\n+----------+-------------------------------------------------+------------------+"
}
```

Maybe a flattened down version of the hex dump might be useful, something like:
```
+--------------------------------------------------------------------------------+
|                                             ↓                                  |
+--------------------------------------------------------------------------------+
|i":"2007-12-03T10:15:30.001Z","ld":"2007-12-3","ldt":"2007-12-03T10:15:30","lt":|
+--------------------------------------------------------------------------------+
```
That could at least be better understood by an end user I reckon, but still has formatting issues for a json based api response.

I would also suggest to make the hex dump a separate optional property on the `JsonReaderException`, to keep error message separate from error context. Again for logging purposes vs (end)user informing.

Still I think it is best to keep the technical hex dump away from my end users, but since it's quite hard to parse a meaningful message out of it, that is currently the only useful thing I can return.

I think that adding the key, that's in scope when reading fails, to the exception would be very helpful. I can then just return with an error message like: 
```unable to read value for key 'X': expected digit, offset: 0x0000012d```

Thoughts?

**Comment by plokhotnyuk** [comment]
@momania thanks for the feedback!

Your flattened down version of the hex dump doesn't work for all possible cases of errors reported by the parser. Malformed bytes, broken surrogate pairs, etc.

Also, your idea to print the key name doesn't work for all arbitrary messages. Let's imagine that some key can appears  1000 times in a request, will it be handy it this case?

**Comment by momania** [comment]
Ah, you've got a point there yes!
I'll give it some more thoughts over the weekend :-)

**Comment by momania** [comment]
Just wanted to leave a comment that I haven't forgotten about this, but also haven't looked into or thought about a way to make the error messages more user friendly.
Maybe, considering jsoniter-scala should stay high performance, it's even an option to have an alternative parser that has very good error handling, and do a second parsing round with that one in case of errors. The performance penalty in an error case usually doesn't matter anyway.

**Comment by jvican** [comment]
I'm personally missing more context-specific information to debug errors reported from jsoniter-scala.
  
Some of these exceptions have thrown from the internal guts of services where I don't know what specific part failed to parse what specific data. Having some location info pointing to the place where `readFrom*` methods are called would help. This could be done both a compile time or at runtime.
  
@momania's suggestion is also interesting because it would allow jsoniter-scala to keep its default of stackless exceptions and only fallback to non-performant parsing whenever it wants. However, I expect it to be really tricky --if not impossible-- to implement because input streams and byte buffers can only be iterated through once, so this would somehow require materializing all of the input first as a string when reading from byte buffers or input streams. If reading from strings, it would be possible.

**Comment by plokhotnyuk** [comment]
@jvican it is possible to switch from stack-less to exception with stack traces using reader and writer configs in runtime:
```
readFromArray[MissingRequiredFields](jsonBytes, ReaderConfig.withThrowReaderExceptionWithStackTrace(isDebug))
```

**Comment by jvican** [comment]
I’m aware of that, but I call readFromArray from many places and I can’t change the reader config (nor I want to carry it around all use sites).

I don’t expect gathering the location info of the readFromArray method would impact negatively performance. We could enrich the exception thrown with the location info just before returning from the read method. Have you ever thought of doing this?

The default of missing key debugging information is IMO what we should strive to change.

**Comment by plokhotnyuk** [comment]
As an option users can wrap all `read...` and `write...` calls by some adapter which can change default configs or catch errors and rethrow them as exceptions with stack traces.

**Comment by jvican** [comment]
What worries me is the default behavior, though. I think we can improve it without compromising performance.
